### PR TITLE
Replace Hadoop-BAM with Disq

### PR DESCRIPTION
**Pull request by tomwhite**
Preview of the changes needed to use Disq.

### CODE REVIEW

**Comment by tomwhite** [comment]
There are currently two failing tests, both of which need fixes in htsjdk.
* CountVariantsSparkIntegrationTest. This is a concurrency issue where VCFCodec (which isn't threadsafe) is being shared by multiple tasks in each Spark executor. The fix is for each task (partition) to have its own VCFCodec. This needs a couple of small changes in htsjdk to make it possible to access the codec's version and header fields so the codec can be recreated. See https://github.com/samtools/htsjdk/pull/1176.
* CpxVariantReInterpreterSparkIntegrationTest. I tracked this down to a problem with the buffer in htsjdk's SeekableBufferedStream. See https://github.com/samtools/htsjdk/pull/1175 for the fix.

**Comment by jamesemery** [comment]
@tomwhite After spending some time searching for this feature for my testing purposes, it would be helpful to expose the NIO adapter toggle directly from the command line in this branch.

**Comment by tomwhite** [comment]
Htsjdk and the Disq snapshot have been updated and now the previously failing tests are passing.

File: src/main/java/org/broadinstitute/hellbender/utils/variant/writers/GVCFBlockMergingIterator.java [hunk 1]
```diff
@@ -0,0 +1,14 @@
+package org.broadinstitute.hellbender.utils.variant.writers;
+
+import htsjdk.variant.variantcontext.VariantContext;
+import org.broadinstitute.hellbender.utils.iterators.PushToPullIterator;
+
+import java.util.*;
+
+public class GVCFBlockMergingIterator extends PushToPullIterator<VariantContext> {
```
Comment (jamesemery) [inline src/main/java/org/broadinstitute/hellbender/utils/variant/writers/GVCFBlockMergingIterator.java]: I would add a comment here making explicit the fact that you are reusing the PushToPullIterator which talks about downsampling in its comments and indeed appears to have that hardwired into its behavior. Or I would make the commenting on PushToPullIterator more generic (referring to downsampling as an example) to avoid confusion.

File: src/main/java/org/broadinstitute/hellbender/utils/variant/writers/GVCFBlockCombiner.java [hunk 2]
```diff
@@ -0,0 +1,220 @@
...
+
+import java.util.ArrayDeque;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Queue;
+
+import static htsjdk.variant.vcf.VCFConstants.MAX_GENOTYPE_QUAL;
+import static org.broadinstitute.hellbender.utils.variant.writers.GVCFWriter.GVCF_BLOCK;
+
+public final class GVCFBlockCombiner implements PushPullTransformer<VariantContext> {
```
Comment (jamesemery) [inline src/main/java/org/broadinstitute/hellbender/utils/variant/writers/GVCFBlockCombiner.java]: Could you actually spin all of the push/pull iterator code into a separate branch? It deserves a separate PR from the rest of the code in this branch.

**Comment by tomwhite** [comment]
Opened #5311 for the push/pull part.

File: src/main/java/org/broadinstitute/hellbender/utils/variant/writers/GVCFBlockMergingIterator.java (hunk 1 above)
Comment (tomwhite) [reply src/main/java/org/broadinstitute/hellbender/utils/variant/writers/GVCFBlockMergingIterator.java]: I've reworked the javadoc in #5311 to address this.

File: src/main/java/org/broadinstitute/hellbender/engine/spark/datasources/ReadsSparkSink.java [hunk 3]
```diff
@@ -185,30 +71,40 @@ public static void writeReads(
...
                                         BucketUtils.makeFilePathAbsolute(referenceFile) :
                                         referenceFile;
-        setHadoopBAMConfigurationProperties(ctx, absoluteOutputFile, absoluteReferenceFile, format);
+      ReadsSparkSource.checkCramReference(ctx, absoluteOutputFile, absoluteReferenceFile);
 
         // The underlying reads are required to be in SAMRecord format in order to be
         // written out, so we convert them to SAMRecord explicitly here. If they're already
         // SAMRecords, this will effectively be a no-op. The SAMRecords will be headerless
         // for efficient serialization.
+        // TODO: add header here
```
[the remark this answers is not in the document]
Comment (tomwhite) [reply src/main/java/org/broadinstitute/hellbender/engine/spark/datasources/ReadsSparkSink.java]: Done

File: src/main/java/org/broadinstitute/hellbender/engine/spark/datasources/ReadsSparkSink.java [hunk 4]
```diff
@@ -217,6 +113,22 @@ public static void writeReads(
         }
     }
 
+    private static void writeReads(
+            final JavaSparkContext ctx, final String outputFile, final String referenceFile, final JavaRDD<SAMRecord> reads,
+            final SAMFileHeader header, final int numReducers, final WriteOption... writeOptions) throws IOException {
+
+        final JavaRDD<SAMRecord> sortedReads = sortSamRecordsToMatchHeader(reads, header, numReducers);
```
[the remark this answers is not in the document]
Comment (tomwhite) [reply src/main/java/org/broadinstitute/hellbender/engine/spark/datasources/ReadsSparkSink.java]: No tests have been removed, and they all pass. Do you think there is another case that needs covering?

We can add logging to Disq for some of these operations.

File: src/main/java/org/broadinstitute/hellbender/engine/spark/datasources/ReadsSparkSource.java [hunk 5]
```diff
@@ -216,69 +162,35 @@ public SAMFileHeader getHeader(final String filePath, final String referencePath
...
-        final Configuration conf = ctx.hadoopConfiguration();
-        conf.set(SAMHeaderReader.VALIDATION_STRINGENCY_PROPERTY, validationStringency.name());
-
-        if (!IOUtils.isCramFileName(inputName)) {
-            // only set the reference for CRAM input
-            conf.unset(CRAMInputFormat.REFERENCE_SOURCE_PATH_PROPERTY);
-        }
-        else {
-            if (null == referenceName) {
+    static String checkCramReference(final JavaSparkContext ctx, final String filePath, final String referencePath) {
```
[the remark this answers is not in the document]
Comment (tomwhite) [reply src/main/java/org/broadinstitute/hellbender/engine/spark/datasources/ReadsSparkSource.java]: Done

File: src/main/java/org/broadinstitute/hellbender/engine/spark/datasources/VariantsSparkSink.java [hunk 6]
```diff
@@ -202,29 +66,22 @@ private static void writeVariantsSingle(
...
 
-        final Configuration conf = ctx.hadoopConfiguration();
-
-        //TODO remove me when https://github.com/broadinstitute/gatk/issues/4303 are fixed
+        //TODO remove me when https://github.com/broadinstitute/gatk/issues/4303 is fixed
         if (outputFile.endsWith(IOUtil.BCF_FILE_EXTENSION) || outputFile.endsWith(IOUtil.BCF_FILE_EXTENSION + ".gz")) {
             throw new UserException.UnimplementedFeature("It is currently not possible to write a BCF file on spark.  See https://github.com/broadinstitute/gatk/issues/4303 for more details .");
         }
-
-        if (outputFile.endsWith(BGZFCodec.DEFAULT_EXTENSION) || outputFile.endsWith(".gz")) {
```
[the remark this answers is not in the document]
Comment (tomwhite) [reply src/main/java/org/broadinstitute/hellbender/engine/spark/datasources/VariantsSparkSink.java]: We do, this case is now handled in Disq. See e.g. https://github.com/disq-bio/disq/blob/master/src/test/java/org/disq_bio/disq/HtsjdkVariantsRddTest.java#L32.

File: src/main/java/org/broadinstitute/hellbender/engine/spark/datasources/VariantsSparkSource.java [hunk 7]
```diff
@@ -65,16 +59,13 @@ public VariantsSparkSource(JavaSparkContext ctx) {
      * @return JavaRDD<VariantContext> of variants from all files.
      */
     public JavaRDD<VariantContext> getParallelVariantContexts(final String vcf, final List<SimpleInterval> intervals) {
-        Configuration conf = new Configuration();
-        conf.setStrings("io.compression.codecs", BGZFEnhancedGzipCodec.class.getCanonicalName(),
```
[the remark this answers is not in the document]
Comment (tomwhite) [reply src/main/java/org/broadinstitute/hellbender/engine/spark/datasources/VariantsSparkSource.java]: Handled by Disq.

File: src/test/java/org/broadinstitute/hellbender/engine/spark/datasources/ReadsSparkSinkUnitTest.java [hunk 8]
```diff
@@ -152,7 +153,7 @@ private void assertSingleShardedWritingWorks(String inputBam, String referenceFi
 
         // check that a splitting bai file is created
         if (IOUtils.isBamFileName(outputPath)) {
-            Assert.assertTrue(Files.exists(IOUtils.getPath(outputPath + SplittingBAMIndexer.OUTPUT_FILE_EXTENSION)));
+            //Assert.assertTrue(Files.exists(IOUtils.getPath(outputPath + SBIIndex.FILE_EXTENSION)));
```
[the remark this answers is not in the document]
Comment (tomwhite) [reply src/test/java/org/broadinstitute/hellbender/engine/spark/datasources/ReadsSparkSinkUnitTest.java]: I opened https://github.com/disq-bio/disq/issues/45 to do this. It shouldn't block this being merged though.

File: build.gradle [hunk 9]
```diff
@@ -53,6 +53,10 @@ repositories {
         url "https://broadinstitute.jfrog.io/broadinstitute/libs-snapshot/" //for htsjdk snapshots
     }
 
+    maven {
+        url "https://oss.sonatype.org/content/repositories/snapshots/" //for disq snapshots
```
[the remark this answers is not in the document]
Comment (tomwhite) [reply build.gradle]: Disq 0.1.0 is now in Maven central, and I've updated this PR to use it.

**Comment by lbergelson** [comment]
Yay!
